pageserver/src/tenant/timeline.rs

```diff
@@ -1809,13 +1819,9 @@ impl Timeline {
 
         if let Some(remote_client) = &self.remote_client {
             for (path, layer_metadata) in layer_paths_to_upload {
-                remote_client
-                    .schedule_layer_file_upload(&path, &layer_metadata)
-                    .context("schedule_layer_file_upload")?;
+                remote_client.schedule_layer_file_upload(&path, &layer_metadata)?;
             }
-            remote_client
-                .schedule_index_upload(&metadata)
-                .context("schedule_layer_file_upload")?;
+            remote_client.schedule_index_upload_for_metadata_update(&metadata)?;
```

Comment:
> `latest_gc_cutoff` might've changed too. 

* if GC had removed a layer during this, `schedule_index_upload_for_file_changes` will work here too
* if GC did not remove anything, `latest_gc_cutoff` loss will not lead to the timetravel bug either: the GC will be able to correctly reconstruct this value on restart

So, it looks like we can have a single method still?

Reply:
> * if GC had removed a layer during this, `schedule_index_upload_for_file_changes` will work here too

Hmm. `schedule_index_for_metadata_update` doesn't currently take a `metadata` argument at all. We could add it there, of course. But then it would be weird if it the function uploaded the index only if the files changed, and ignored the metadata otherwise. We could define it so that it performs the upload, if there have been changes to the metadata, or to the files. Overall, though, I think it's more clear to have separate functions.

That said, I'm not totally happy with this interface either. I think two functions feels more clear than one, but some sort of refactoring probably would be in order here. One radical idea is to get rid of the upload queue and all the `schedule_*` calls altogether, and instead add more state to the individual Layers to keep track which ones exist remotely and which exist locally, and which ones are present in the remote index file. Then have some code that scans the layer map for those states, and does the right things to make progress. Or something like that...